### ISSUE TITLE
Unnecessary canonicalization?

### ISSUE THREAD

**Issue by modulovalue** [bishoksan/DFTA#3]
Consider:

https://github.com/bishoksan/DFTA/blob/2e2c09c89869c6dd85f4b76e25a2f5b177f349d1/src/dfta/DeterminiserOpt.java#L225-L231

I'm wondering and I wanted to ask: what's the purpose of the `qdnames` hashmap here? `qdnames` isn't mutated once it has been built and so It looks to me like `qdnames` acts like the identity function for elements of `qd`, so it seems like the `qdnames` hashmap could be removed entirely?

**Comment by jpgallagher** [comment]
Thanks for noticing this. It's quite some time since I last looked at this code and I must say qdnames looks redundant when I look at it now.  So all occurrences of qdnames.get(q0) could just be replaced by q0. Perhaps when I wrote the code I was anticipating building some other representation of a set of String.  However I cannot reconstruct now why this map was introduced. Feel free to change the code and note if you get any different results in the number of transitions of the output dfta.  For the time being I'll just leave the code as it is.

**Comment by modulovalue** [comment]
Hey @jpgallagher

I just wanted to let you know that I've tried to refactor and clean up DFTA while reading your paper.

It can be found in the master branch of my fork here: https://github.com/modulovalue/DFTA

In particular, these two files (and the parser) contain everything that's needed: https://github.com/modulovalue/DFTA/blob/master/src/dfta/Determiniser.java and https://github.com/modulovalue/DFTA/blob/master/src/dfta/Driver.java

I've verified that determinised trees are identical to the trees produced by your version of the repo here: https://github.com/modulovalue/DFTA/tree/orig

Unfortunately, I had to remove a couple of important parts (but I plan to add them back in the future) which are:
- dontCares (In practice, and according to the benchmarks in your paper, they appear not to add much value)
- anything UI related (it didn't appear to add much value)
- inclusion testing
- the Carrasco minimization implementation

If you'd like to, we can merge the changes into the main repo, but no worries if you'd rather not do that and you'd prefer to keep the original implementation.

**Comment by jpgallagher** [comment]
Hi @modulovalue,

Looks nice, but I would need to take a longer look at your code to decide whether to merge.  Also, although don't cares do not give much performance benefit, they might offer advantages when using the DFTAs, e.g. when representing them using BDDs as in our previous work (2005). So I'd like to keep that.

Is there any performance gain with your version?

I'm curious about your interest in DFTAs.  Have you some application in mind, are you planning to publish something, or is it just for fun :-) - a great way to spend Christmas Day ;-) ?  If you really need performance, perhaps recoding the algorithm in C/C++ would be the way to go.

I have retired now but still doing some research. @bishoksan has also moved on to other things.  But if you're interested in collaborating on some further work on this stuff let us know.  Did you notice I committed some further code I worked on later, using a SAT solver to perform some operations in the algorithm?  This did not yield any interesting results but I still think it might be worth following up.

Regards
John

**Comment by modulovalue** [comment]
> Also, although don't cares do not give much performance benefit, they might offer advantages when using the DFTAs, e.g. when representing them using BDDs as in our previous work (2005). So I'd like to keep that.

I agree, it feels like an important idea. My main motivation for removing them (for now) was to make it easier for myself to focus on the core idea of your paper.

> Did you notice I committed some further code I worked on later, using a SAT solver to perform some operations in the algorithm? 

I did, and thank you for doing that!

My BDD (and SAT) knowledge isn't extensive, so I can't quite judge the implications of, e.g., "dontCares" when it comes to using those different approaches yet, but thank you for making the connection. I will keep that in mind and I do plan to eventually look deeper into those topics.

> Is there any performance gain with your version?

So far, I've tried not to deviate too much from your original implementation, but I did manage to improve the performance of, e.g., determinising A1003 from 49 seconds with your version, to 30 seconds with an implementation written in Dart.

In any case, there are two performance improvements that your implementation could benefit from:

- function symbols and states can be canonicalized to integer indexes, which allows sets of them to be stored as BitSets. Sets of powersets can then be represented as Sets of BitSets, and Maps can be represented as plain old Arrays. This avoids a lot of hashmap/hashset-related overhead.

(That investigation has lead to many performance optimizations for a BitSet implementation written in Dart https://github.com/isoos/bit_array/issues/24. A hashCode cache is an important one once sets of powersets are represented as sets of BitSets)

- the delta set construction for constant function symbols can be inlined into the index so that they are calculated only once:

The exact change can be found in this commit: https://github.com/modulovalue/DFTA/commit/878d90c8fe2997b6100bee190f74117a0a5e7e42	

---

> I'm curious about your interest in DFTAs. Have you some application in mind

I do! I'll send you an email.
